### ISSUE TITLE
Support make jobserver protocol

### ISSUE THREAD

**Issue by rossburton** [michaelforney/samurai#71]
GNU Make has a neat feature called the jobserver protocol, where the top-level Make can allocate a specific number of job slots, and child makes can take slots to do work in. This was designed to stop the parallelisation problem where a top-level `make -j10` may potentially spawn 10 separate sub-makes all with `-j10` so there's now 100 parallel jobs.

However, it's also useful for resource control is systems which built multiple pieces of software at once.  For example, Bitbake can build N different pieces of software at once, and each of those is passed a `-jM` flag. If each of these N tasks is compiling then thats's N*M jobs so you don't want N or M to be too high, but if only 1 of N is building then you want M to be high.

With the job server protocol there are N slots in total for all sub makes, so you can control the resource utilisation more accurately.  If Samurai also supported the jobserver protocol instead of just `-j` then it could join in the resource pooling and builds can be more efficient.

The canonical implementation of jobserver is in the GNU Make sources. The documentation is pretty vague as it is considered an implementation detail, but it hasn't changed in 20 years...  The prototype BitBake support for a master jobserver queue is at http://git.yoctoproject.org/cgit.cgi/poky-contrib/commit/?h=rpurdie/wipqueue4&id=d66a327fb6189db5de8bc489859235dcba306237

**Comment by rossburton** [comment]
cc @rpurdie

**Comment by bonzini** [comment]
Hi, I have done a quick and dirty implementation (less than 200 lines of code) but it uses pthreads. Should I submit a pull request or should I first try to rework it without threads?

**Comment by michaelforney** [comment]
@bonzini, I am interested in taking a look at what you've got, but I'd definitely prefer an implementation without threads.

**Comment by eli-schwartz** [comment]
BTW since https://git.savannah.gnu.org/cgit/make.git/commit/?id=7ad2593b2d2bb5b9332f4444d8bf93ac6f958bc6 GNU make supports named pipes as an alternative to anonymous pipes.

Once it actually gets into a release that should be pretty nifty.

**Comment by mid-kid** [comment]
I feel it's worth noting that gcc picks up on this jobserver when using `gcc -flto` for LTO builds, and as such is very useful for speeding that up. Using `-flto=n` to specify the amount of LTO linking jobs isn't great when it comes to projects linking multiple programs concurrently.

**Comment by bonzini** [comment]
> @bonzini, I am interested in taking a look at what you've got, but I'd definitely prefer an implementation without threads.

I opened a new PR in #104 now that someone reminded me of it. It does use threads, which is almost mandatory because the GNU Make protocol requires a blocking file descriptor. But the implementation is quite nifty if you ask me. :)

**Comment by michaelforney** [comment]
> It does use threads, which is almost mandatory because the GNU Make protocol requires a blocking file descriptor.

Do you know why the protocol uses blocking file descriptors? That seems like a weird choice to me, and as you say, require the use of threads to implement properly.

**Comment by bonzini** [comment]
Well it's an old protocol... The idea was that when waiting on a blocking read you had nothing to do anyway; and with some care to avoid races, the read could be interrupted by SIGCHLD (the only other event that Make processed). Of course these days it's a bit inconvenient if you have any kind of event loop; but the solution with the internal pipe, despite the usage of threads, actually simplifies noticeably the handling of the "free" token (which is important to avoid starvation).

**Comment by haampie** [comment]
I just learned about samurai and was curious if GNU Jobserver support was implemented here, now that ninja has it. Stumbling on this thread I thought it'd be good to clear a few things up.

GNU Make implemented the Jobserver protocol in 1999. Linux only added Native POSIX Thread Library support in 2003. So GNU make was necessarily single-threaded, and still is. It also does not use blocking reads.

It works with a pselect based event loop, where it waits on two things: the jobserver fd to become readable, or SIGCHLD through an interrupt:

```c

/* block SIGCHLD everywhere, except when waiting in pselect below */
sigemptyset(&empty_mask);
sigemptyset(&block_mask);
sigaddset(&block_mask, SIGCHLD);
sigprocmask(SIG_BLOCK, &block_mask, NULL);

/* this is a simplified "acquire" function */
while (1) {
    int r = pselect(job_fds[0] + 1, &readfds, NULL, NULL, NULL, &empty_mask);  // this blocks

    if (r < 0 && errno == EINTR) {  // interrupt means SIGCHLD
        reap_children();
        continue;
    }

    char token;
    if (read(job_fds[0], &token, 1) == 1) { // pipe is readable, this read is non-blocking
        return;
    } else {
        continue;  // EAGAIN: our process was too late, get stuck in pselect again
    }
}
```

**Comment by michaelforney** [comment]
Thanks @haampie for providing some additional context here.

With this approach, the main concern is the race condition described by @bonzini in https://github.com/michaelforney/samurai/pull/94#discussion_r1572716430. What happens when a single job token becomes available and multiple sub-makes are waiting for a new job? Only one of the will get it, so won't the `read()` block for the other?

In the code snippet, you have a case handling `EAGAIN` but I don't think you'll get that unless the file descriptor is non-blocking.

**Comment by bonzini** [comment]
> GNU Make implemented the Jobserver protocol in 1999. Linux only added Native POSIX Thread Library support in 2003. So GNU make was necessarily single-threaded, and still is. It also does not use blocking reads.

GNU make's original implementation used dup() + closing the file descriptor in a signal handler for SIGCHLD. The implementation with pselect() was a necessary workaround that Make added due to incorrect programs that set the pipe to non-blocking. However, a single-threaded implementation with pselect() is unfortunately buggy. The only way to make it robust for both blocking and non-blocking file descriptors is to make it multi-threaded.

My implementation of the jobserver protocol was stuck on the first part that fixed the implementation of SIGTERM in samurai (which was a prerequisite for the jobserver work). I'll try to get back to it.

**Comment by haampie** [comment]
If you worry about sub-processes making the shared fd blocking, notice that upstream ninja only implements the FIFO-based jobserver where you don't have that problem. I wouldn't implement the old pipe version of the protocol this day and age.
